### PR TITLE
Not adding prefixes to names after use statements

### LINKED ISSUE #456: Use directive ignored for prefixes
## Bug report

| Question    | Answer
| ------------| ---------------
| PHP-Scoper version | 0.14.0
| PHP version | 7.4.15
| Platform with version | Debian testing
| Github Repo | Not public

Hi,
I'm trying to use php-scoper on the Google SDK and I am getting prefixes for classes that have already been brought into scope with the `use` directive. I tried shortening it a bit, but here's my example:

Source file:
```php
<?php
namespace Google\AccessToken;

use Google\Auth\HttpHandler\HttpHandlerFactory;
use Google\Client;
use GuzzleHttp\ClientInterface;
use GuzzleHttp\Psr7;
use GuzzleHttp\Psr7\Request;

/**
 * Wrapper around Google Access Tokens which provides convenience functions
 *
 */
class Revoke
{
  /**
   * @var ClientInterface The http client
   */
  private $http;

  /**
   * Instantiates the class, but does not initiate the login flow, leaving it
   * to the discretion of the caller.
   */
  public function __construct(ClientInterface $http = null)
  {
    $this->http = $http;
  }
}
```

Output:
```php
<?php
namespace Sgdg\Vendor\Google\AccessToken;

use Sgdg\Vendor\Google\Auth\HttpHandler\HttpHandlerFactory;
use Sgdg\Vendor\Google\Client;
use Sgdg\Vendor\GuzzleHttp\ClientInterface;
use Sgdg\Vendor\GuzzleHttp\Psr7;
use Sgdg\Vendor\GuzzleHttp\Psr7\Request;
/**
 * Wrapper around Google Access Tokens which provides convenience functions
 *
 */
class Revoke
{
    /**
     * @var ClientInterface The http client
     */
    private $http;
    /**
     * Instantiates the class, but does not initiate the login flow, leaving it
     * to the discretion of the caller.
     */
    public function __construct(\Sgdg\Vendor\GuzzleHttp\ClientInterface $http = null)
    {
        $this->http = $http;
    }
}
```

As you can see, in the constructor, the `ClientInterface` class is prefixed, but AFAIK that is not necessary as it has already been brought into scope with the `use` directive.

### PR DESCRIPTION
**Pull request by marekdedic**
Closes #456 

I am new to the codebase and it isn't exactly simple, so please check if this breaks anything...

### CODE REVIEW

**Comment by marekdedic** [comment]
Ok, so I'm going through the failing unit tests and fixing them, but there are a few cases that need comments:
- `e2e_026`: I think it was broken before, but the test didn't catch it - the `use const` statement wasn't prefixed (but should've been IMO)

**Comment by marekdedic** [comment]
I think this is ready for review. Please review the changes. I'd focus on stuff around whitelisting as I am still a little fuzzy about the details of how it's supposed to work. And also e2e test 32 where I had to put in a special workaround - I'm not really sure whether there isn't done bigger underlying issue I'm missing completely or is it's just caused by using php-scoper on itself.


BTW all the tests really helped to get this into shape, thanks for being diligent about them!

**Comment by theofidry** [comment]
Sorry for the tardiness @marekdedic, I'll try to check it out ASAP

**Comment by marekdedic** [comment]
Thanks, don't worry, I know what it's like with volunteer-run projects :slightly_smiling_face:

File: src/PhpParser/NodeVisitor/NameStmtPrefixer.php [hunk 1]
```diff
@@ -204,4 +226,14 @@ private function prefixName(Name $name): Node
             $resolvedName->getAttributes()
         );
     }
+
+    private static function array_starts_with($arr, $prefix): bool
+    {
+        for ($i = 0; $i < count($prefix); ++$i) {
```
Comment (theofidry) [inline src/PhpParser/NodeVisitor/NameStmtPrefixer.php:232]: be aware that the count here is executed every single time which is not ideal

File: src/PhpParser/NodeVisitor/NameStmtPrefixer.php [hunk 2]
```diff
@@ -146,6 +155,19 @@ private function prefixName(Name $name): Node
...
+        if (
+            $useStatement !== null and
+            self::array_starts_with($resolvedName->parts, $useStatement->parts) and
+            !($parentNode instanceof ConstFetch and ($this->whitelist->isGlobalWhitelistedConstant($resolvedName->toString()) or $this->whitelist->isSymbolWhitelisted($resolvedName->toString(), true))) and
+            !($useStatement->getAttribute('parent') and $useStatement->getAttribute('parent')->alias !== null and $this->whitelist->isSymbolWhitelisted($useStatement->toString())) and
+            !($name instanceof FullyQualified) and
+            $resolvedName->parts !== ['Isolated', 'Symfony', 'Component', 'Finder', 'Finder']
+        ) {
+            return $name;
+        }
```
Comment (theofidry) [inline src/PhpParser/NodeVisitor/NameStmtPrefixer.php:169]: There is several things I'm not too fond of here:

- the `and` (see https://github.com/kalessil/phpinspectionsea/blob/master/docs/control-flow.md#non-optimal-if-conditions)
- the long condition

the later though is hardly specific from your addition I'm already unpleased with the existing code in that regard. I'd just change a bit the indentation to make it a bit easier to follow:

```suggestion

        if (
            $useStatement !== null
            && !($name instanceof FullyQualified)
            && self::array_starts_with($resolvedName->parts, $useStatement->parts)
            && !(
                $parentNode instanceof ConstFetch
                && (
                    $this->whitelist->isGlobalWhitelistedConstant($resolvedName->toString())
                    || $this->whitelist->isSymbolWhitelisted($resolvedName->toString(), true)
                )
            )
            && !(
                $useStatement->getAttribute('parent')
                && $useStatement->getAttribute('parent')->alias !== null
                && $this->whitelist->isSymbolWhitelisted($useStatement->toString())
            )
            && $resolvedName->parts !== ['Isolated', 'Symfony', 'Component', 'Finder', 'Finder']
        ) {
            return $name;
        }
```

**Review by theofidry — APPROVED** [review]
I did not bother with a Blackfire profile but I can't see any notable speed difference so LGTM, thanks a lot for the impressive work!

>I'd focus on stuff around whitelisting as I am still a little fuzzy about the details of how it's supposed to work.

Regarding the whitelisting it's something I would like to change as the terminology is both deceptive and inappropriate (without going too far into diversity stuff I am not too fond of the whitelist/blacklist terminology).

Ideally I would like to come with terms that can properly express that in one case the symbols are left unchanged (e.g. for "whitelist-files" the files are not changed at all) and in the other they are still scoped but an alias is added to allow using its non-scoped symbol (more details on this in https://github.com/humbug/php-scoper#implementation-insights)

>And also e2e test 32 where I had to put in a special workaround - I'm not really sure whether there isn't done bigger underlying issue I'm missing completely or is it's just caused by using php-scoper on itself.

I am a bit confused by that one as well. In case this is unclear:

- PHP-Scoper ships with a Finder which is scoped
- Since PHP-Scoper loads directly the PHP config file, although there is other ways, in order to leverage the Finder without requiring the user to autoload special code PHP-Scoper exposes its finder (see `bin/php-scoper`:)

```php
// Exposes the finder used by PHP-Scoper PHAR to allow its usage in the configuration file.
if (false === class_exists(IsolatedFinder::class)) {
    class_alias(Finder::class, IsolatedFinder::class);
}
```

However the above would still be scoped and to avoid that I am relying on a patcher in the config:

```
        static function (string $filePath, string $prefix, string $contents): string {
            if ('bin/php-scoper' !== $filePath) {
                return $contents;
            }

            return str_replace(
                '\\'.$prefix.'\Isolated\Symfony\Component\Finder\Finder::class',
                '\Isolated\Symfony\Component\Finder\Finder::class',
                $contents
            );
        },
```

The fixture 032 is to make sure this process works, i.e. that the isolated finder is exposed and usable.

>BTW all the tests really helped to get this into shape, thanks for being diligent about them!

Glad they helped out, they are annoying to write and a bit slow but hopefully they pay for themselves

File: src/PhpParser/NodeVisitor/NameStmtPrefixer.php (hunk 1 above)
Comment (marekdedic) [reply src/PhpParser/NodeVisitor/NameStmtPrefixer.php]: Ha, forgot I was in one of the languages that does this, thanks! :)

**Comment by marekdedic** [comment]
I agree that the term "whitelisted" can be really confusing for this use. If I may, I'd suggest ignored/excluded for files that are completely untouched. For the other use, maybe something like "add-alias-to-original"?

I fixed both your suggestions, thank you for them. Is there anything more to do on my part? (e.g. regarding the Finder or whitelisting?)

**Comment by theofidry** [comment]
>I'd suggest ignored/excluded for files that are completely untouched. For the other use, maybe something like "add-alias-to-original"?

I think those are good suggestions

>I fixed both your suggestions, thank you for them. Is there anything more to do on my part? (e.g. regarding the Finder or whitelisting?)

I think the Finder can wait; it's all good I just wanted to do a patch release before merging this :)

**Comment by theofidry** [comment]
Thanks a lot @marekdedic

**Comment by TomasVotruba** [comment]
Thank you @marekdedic :+1:

**Comment by TomasVotruba** [comment]
@theofidry Hi Theo, this feature would be very helpful with solving https://github.com/humbug/php-scoper/issues/440#issuecomment-795160132

Too bad patch release was before this and not after :D
Any ETA on phar release?

**Comment by TomasVotruba** [comment]
@theofidry Thank you for such a fast action :clap: 
I'm gonna try it on Rector right now

**Comment by marekdedic** [comment]
Hi @TomasVotruba, I am a bit lost in your code and cannot reproduce the issue. Could you please narrow it down to some minimal reproducible example?

Thanks.

**Pull request merged**